### PR TITLE
MulticlassNms/MatrixNms: ngraph op support FP16 output

### PR DESCRIPTION
**Pull request by luo-cheng2021**
### Details:
 - *recover overwritten code in the #8520, also resolve the MulticlassNms/MatrixNms problem mentioned [here](https://github.com/openvinotoolkit/openvino/pull/8394#issuecomment-958720437)*
 - *op support FP16 output*

### Tickets:
 - *74710*

### CODE REVIEW

File: src/inference/src/cnn_network_ngraph_impl.cpp [hunk 1]
```diff
@@ -439,8 +439,8 @@ void CNNNetworkNGraphImpl::reshape(const std::map<std::string, ngraph::PartialSh
                 manager.register_pass<ov::pass::RemoveConcatZeroDimInput>();
                 manager.register_pass<ov::pass::RemoveMultiSubGraphOpDanglingParams>();
                 manager.register_pass<::ngraph::pass::ConvertNMS5ToLegacyMatcher>(false);
-                manager.register_pass<::ngraph::pass::ConvertMulticlassNmsToMulticlassNmsIE>();
-                manager.register_pass<::ngraph::pass::ConvertMatrixNmsToMatrixNmsIE>();
+                manager.register_pass<::ngraph::pass::ConvertMulticlassNmsToMulticlassNmsIE>(false);
+                manager.register_pass<::ngraph::pass::ConvertMatrixNmsToMatrixNmsIE>(false);
```
Comment (mvafin) [inline src/inference/src/cnn_network_ngraph_impl.cpp:443]: Why these changes needed?

Comment (luo-cheng2021) [reply src/inference/src/cnn_network_ngraph_impl.cpp:443]: The transformation will not insert 'convert' node to convert 'i64' to 'i32' with the parameter 'false'. If add 'convert' node, the network output name in 'reshape' will change to 'XXX_Nms_0/convert.1'. In CPU plugin when nms node followed by 'Result' node the transformation will be skipped and the network output name will be 'XXX_Nms_0', this is conflict with the previous one.
There are some ways to resolve the problem:
1, (currently used) Tell the transformation not to add 'convert' node in the caller 'reshape', followed the original [NonMaxSuppression design](https://github.com/luo-cheng2021/openvino/blob/9a5b91185679d67b83a6841603454006dc9f1d09/src/inference/src/cnn_network_ngraph_impl.cpp#L440)
2, Always add 'convert' node even if the transformation is skipped. There is a little concern: calling the transformation in CPU plugin may be removed someday.

File: src/core/src/op/util/nms_base.cpp [hunk 2]
```diff
@@ -128,7 +132,7 @@ void ov::op::util::NmsBase::validate_and_infer_types() {
 
     // 'selected_outputs' have the following format:
     //      [number of selected boxes, [class_id, box_score, xmin, ymin, xmax, ymax]]
-    set_output_type(0, element::f32, {first_dim_shape, 6});
+    set_output_type(0, get_input_element_type(0), {first_dim_shape, 6});
```
Comment (rkazants) [inline src/core/src/op/util/nms_base.cpp:135]: please correct the specification. Now the logic contradicts the specification.
@jane-intel, fyi.

**Review by rkazants — CHANGES_REQUESTED** [review]
Please correct the specification to align it with new logic

File: src/core/src/op/util/nms_base.cpp (hunk 2 above)
Comment (luo-cheng2021) [reply src/core/src/op/util/nms_base.cpp:135]: Updated.

**Comment by luo-cheng2021** [comment]
> Please correct the specification to align it with new logic

Updated.

**Review by mvafin — APPROVED** [review]

File: docs/template_plugin/tests/functional/op_reference/matrix_nms.cpp [hunk 3]
```diff
@@ -461,11 +461,11 @@ std::vector<MatrixNmsParams> generateParams() {
 
 std::vector<MatrixNmsParams> generateCombinedParams() {
     const std::vector<std::vector<MatrixNmsParams>> generatedParams {
-        generateParams<element::Type_t::bf16, element::Type_t::f32, element::Type_t::i32>(),
-        generateParams<element::Type_t::f16, element::Type_t::f32, element::Type_t::i32>(),
+        generateParams<element::Type_t::bf16, element::Type_t::bf16, element::Type_t::i32>(),
```
Comment (ilyachur) [inline docs/template_plugin/tests/functional/op_reference/matrix_nms.cpp:464]: Should we keep old test cases and add new tests?

File: src/core/reference/src/runtime/reference/utils/nms_common.cpp [hunk 4]
```diff
@@ -18,11 +18,29 @@ void nms_common_postprocessing(void* prois,
...
+        float16* ptr = static_cast<float16*>(prois);
+        for (auto i = 0; i < total_num * 6; ++i) {
+            ptr[i] = float16(selected_outputs[i]);
+        }
+    } break;
+    case element::Type_t::f32: {
+        float* ptr = static_cast<float*>(prois);
+        memcpy(ptr, selected_outputs.data(), total_num * sizeof(float) * 6);
+    } break;
+    default:;
```
Comment (ilyachur) [inline src/core/reference/src/runtime/reference/utils/nms_common.cpp:42]: Should we prohibit unsupported element types?

Comment (luo-cheng2021) [reply src/core/reference/src/runtime/reference/utils/nms_common.cpp]: Updated.

File: docs/template_plugin/tests/functional/op_reference/matrix_nms.cpp (hunk 3 above)
Comment (luo-cheng2021) [reply docs/template_plugin/tests/functional/op_reference/matrix_nms.cpp:464]: The second type stands for the output type. Since the output type follows the input type and the old case will not  be correct and no needed any more.

**Review by jane-intel — APPROVED** [review]

File: src/core/src/op/util/nms_base.cpp [hunk 5]
```diff
@@ -58,6 +58,10 @@ void ov::op::util::NmsBase::validate() {
                           is_float_type_admissible(get_input_element_type(1)),
                           "Expected bf16, fp16 or fp32 as element type for the 'scores' input.");
 
+    NODE_VALIDATION_CHECK(this,
+                          get_input_element_type(0) == get_input_element_type(1),
+                          "Expected 'boxes', 'scores' type is same.");
+
```
[the remark this answers is not in the document]
Comment (luo-cheng2021) [reply src/core/src/op/util/nms_base.cpp]: Done.

**Review by rkazants — APPROVED** [review]

**Review by ilyachur — APPROVED** [review]
